### PR TITLE
Add multi modal RCNN model

### PR DESCRIPTION
**Pull request by giovanni-corvi**
Add multi modal RCNN model and tweak data loading to allow description in dataset

### CODE REVIEW

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+
+PROCESSED_IMAGE_SIZE = (768, 1024)
+
+
+class TextFeatureExtractor(torch.nn.Module):
+    def __init__(self,  output_shape: int, expected_output_shape: tuple, fpn_keys: list[str], device: str, interpolation_mode: str = 'bilinear'):
+        """
+        we assume that the output_shape has shape flattened shape K (make sure)
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py]: Are you sure about the "make sure"?

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        we assume that the expected_output_shape has shape (W, X, X) [does it work with w, x, y??]
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py]: Do you already know whether it works with `w,x,y`?

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        self.interpolation_mode = interpolation_mode
+
+        self.device = device
+
+        self.remapper = nn.Sequential(
+            torch.nn.Linear(output_shape, math.prod(expected_output_shape), device=device),
+            nn.LeakyReLU()
+        )
+
+        num_fpn_layers = len(self.fpn_keys)
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py:58]: Can you add some comments on these lines, please?

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+                                self.backbone._square_pad // largest_stride if self.backbone._square_pad else PROCESSED_IMAGE_SIZE[1] // largest_stride)
+
+        self.feature_dropout = nn.Dropout(feature_dropout_rate)
+        self.feature_noise_std = feature_noise_std
+        # TODO: TextFeatureExtractor as input argument
+        self.text_encoder = TextFeatureExtractor(
+            output_shape=384,
+            expected_output_shape=(channels, int(smallest_feature_size[0]), int(smallest_feature_size[1])), 
+            fpn_keys=self.backbone._out_features,
+            device="cuda")
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py]:
```suggestion
            device=self.device)
```

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+        feature_shapes = {k: v.shape for k, v in features.items()}
+        text_features = self.text_encoder([batched_input['text'] for batched_input in batched_inputs], feature_shapes)
+
+        for k in features:
+            # Dropout features
+            features[k] = self.feature_dropout(features[k])
+
+            # Add noise
+            features[k] = features[k] + torch.randn_like(features[k]) * self.feature_noise_std
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py]: Could you please add some extra comments on why you do the feature dropout and add the noise?

**Review by diogo-sr — CHANGES_REQUESTED** [review]

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py (hunk 2 above)
Comment (giovanni-corvi) [reply detectron2/modeling/meta_arch/multi_modal_rcnn.py]: yup

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py (hunk 1 above)
Comment (giovanni-corvi) [reply detectron2/modeling/meta_arch/multi_modal_rcnn.py]: yeah but it has to be hardcoded for now (won't be changed unless we change the SentenceTransformer model)

File: detectron2/modeling/meta_arch/multi_modal_rcnn.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
+"""
+Draft implementation of an extension to the GeneralizedRCNN model to enable it to process both images and textual descriptions as input.
+
+The are plenty of TODOs/improvements that could be explored, including:
+    - Introduce a "FusionModel" to handle the integration/combination of visual and text features.
+      Essentially instead of `features = features + text_features` we would have `features = self.fusion_model(features, text_features)`
+      This is because simply adding the two features is probably too simplistic. The fusion model could be a more traditional network that
+      first concatenates than combines the features or more complex (e.g. gating, attention, regularization)
```
Comment (diogo-sr) [inline detectron2/modeling/meta_arch/multi_modal_rcnn.py]:
```suggestion
      first concatenates and then combines the features or more complex (e.g. gating, attention, regularization)
```
